File: LeetCode/1930. Unique Length-3 Palindromic Subsequences/solution.py
# class Solution:
#     def countPalindromicSubsequence(self, s):
#         res = 0
#         for c in string.ascii_lowercase:
#             i, j = s.find(c), s.rfind(c)
#             if i > -1:
#                 res += len(set(s[i + 1: j]))
#         return res        

class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        n = len(s)
        chars = [chr(x+ord('a')) for x in range(26)]
        left, right = {ch: n for ch in chars}, {ch: -1 for ch in chars}
        res = defaultdict(set)
        for i in range(n):
            right[s[i]] = i  
        for i in range(n)[::-1]:
            left[s[i]] = i 
        for i, ch in enumerate(s):
            for ph in chars:
                if left[ph] < i and right[ph] > i:
                    res[ch].add(ph)
        return sum(len(v) for k, v in res.items())

File: LeetCode/1930. Unique Length-3 Palindromic Subsequences/solution.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        std::vector<int> right(26,0), left(26,0);
        for (auto ch : s)
            right[ch-'a']++;
        vector<bool> seen(26*26,false);
        for (auto ch : s){
            int j = ch-'a';
            right[j]--;
            for (int i = 0; i < 26; i++)
                if (left[i] && right[i])
                    seen[i*26+j] = true;
            left[j]++;
        }
        int res = 0;
        for (auto x : seen)
            res += x;
        return res;
    }
};